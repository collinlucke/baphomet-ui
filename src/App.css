/* TODO: Figure out what to do with this */
