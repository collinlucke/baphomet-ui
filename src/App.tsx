import './styling/index.css';
import { Outlet } from 'react-router-dom';
import { Heading } from './components/shared/Heading';
import { Main } from '@collinlucke/phantomartist';
import { ErrorBoundary } from './components/shared/ErrorBoundary';
import { useState } from 'react';
import { ShowHeadingContext } from './contexts';
import { ErrorContext } from './contexts';
import { AuthenticatedContext } from './contexts';
import { CustomErrorTypes } from './CustomTypes.types';

function App() {
  const [showHeading, setShowHeading] = useState(true);
  const [error, setError] = useState<CustomErrorTypes | undefined>(undefined);
  const [authenticated, setAuthenticated] = useState(false);

  return (
    <div>
      <AuthenticatedContext.Provider
        value={{ authenticated, setAuthenticated }}
      >
        <ShowHeadingContext.Provider value={{ showHeading, setShowHeading }}>
          <ErrorContext.Provider
            value={{
              error,
              setError: setError as (error: {} | undefined) => void
            }}
          >
            {showHeading && <Heading />}
            <Main>
              <Outlet />
            </Main>
            {error && <ErrorBoundary />}
          </ErrorContext.Provider>
        </ShowHeadingContext.Provider>
      </AuthenticatedContext.Provider>
    </div>
  );
}

export default App;
