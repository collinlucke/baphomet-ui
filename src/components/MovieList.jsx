import { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';

const Movie = ({ movie, deleteMovie }) => {
  const { title, year, rated, _id } = movie;
  const tdClassName = 'p-4 align-middle [&amp;:has([role=checkbox])]:pr-0';

  const deleteMovieHandler = () => {
    deleteMovie(_id);
  };

  return (
    <tr className="border-b transition-color hover:bg-muted/50 data-[state=selected:bg-muted">
      <td className={tdClassName}>{title}</td>
      <td className={tdClassName}>{year}</td>
      <td className={tdClassName}>{rated}</td>
      <td>
        <div className="flex gap-2">
          <Link
            to={`/edit/${_id}`}
            className="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-slate-100 h-9 rounded-md px-3"
          >
            Edit
          </Link>
          <button
            color="red"
            type="button"
            onClick={deleteMovieHandler}
            className="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-slate-100 hover:text-accent-foreground h-9 rounded-md px-3"
          >
            Delete
          </button>
        </div>
      </td>
    </tr>
  );
};

const MovieList = () => {
  const [movies, setMovies] = useState([]);

  useEffect(() => {
    const getMovies = async () => {
      try {
        const response = await fetch(`http://localhost:5050/movies`);
        const movies = await response.json();
        setMovies(movies);
      } catch (err) {
        const message = `An error occurred: ${err.message}`;
        console.error(message);
        return;
      }
    };
    getMovies();
  }, [movies.length]);

  const deleteMovie = async id => {
    try {
      await fetch(`http://localhost:5050/movies/${id}`, {
        method: 'DELETE'
      });
      const newMovies = movies.filter(mov => {
        return mov._id !== id;
      });
      setMovies(newMovies);
    } catch (err) {
      const message = `An error occurred: ${err.message}`;
      console.error(message);
      return;
    }
  };

  const movieList = () => {
    return movies.map(mov => {
      return <Movie movie={mov} deleteMovie={deleteMovie} key={mov._id} />;
    });
  };

  return (
    <>
      <h3 className="text-lg font-semibold p-4">Movies</h3>
      <div className="border rounded-lg overflow-hidden">
        <div className="relative w-full overflow-auto">
          <table className="w-full caption-bottom text-sm">
            <thead className="[&amp;_tr]:border-b">
              <tr className="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted">
                <th className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">
                  Title
                </th>
                <th className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">
                  Year
                </th>
                <th className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">
                  Rated
                </th>
                <th className="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">
                  Action
                </th>
              </tr>
            </thead>
            <tbody className="[&amp;_tr:last-child]:border-0">
              {movieList()}
            </tbody>
          </table>
        </div>
      </div>
    </>
  );
};

export default MovieList;
