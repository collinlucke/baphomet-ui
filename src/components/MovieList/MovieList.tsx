import { Button, List, Search } from '@collinlucke/phantomartist';
import { MovieListItem } from './MovieListItem';
import { ChangeEvent, FormEventHandler } from 'react';
import { useNavigate } from 'react-router-dom';
import { useScreenSize } from '../../hooks/useScreenSize';

type Movie = {
  id: string;
  title: string;
  releaseDate?: string;
  rated?: string;
  poster?: string;
};

type MovieData = {
  movieData: {
    allMovies: Movie[];
  };
  searchTerm?: string;
  resultsCount?: number;

  onSearch?: React.FormEventHandler<HTMLFormElement>;
  setSearchTerm?: (e: ChangeEvent<HTMLInputElement>) => void;
  openDeleteModal?: ({ id, title }: { id: string; title: string }) => void;
};

export const MovieList: React.FC<MovieData> = ({
  movieData,
  searchTerm,
  onSearch,
  setSearchTerm,
  openDeleteModal
}) => {
  const navigate = useNavigate();
  const screenSize = useScreenSize();

  const onSearchHandler: FormEventHandler<HTMLFormElement> = e => {
    onSearch?.(e);
  };

  const setSearchTermHandler = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchTerm?.(e);
  };

  const navigateToArena = () => {
    navigate('/arena');
  };

  return (
    <List>
      <Search
        onSearch={onSearchHandler}
        searchTerm={searchTerm}
        searchLabel="Search Movies"
        setSearchTerm={setSearchTermHandler}
        resultsCount={movieData.allMovies.length}
        buttonSize={screenSize}
        inputSize={screenSize}
        useSearchButton={false}
      />
      {movieData.allMovies.length ? (
        movieData.allMovies.map(mov => (
          <MovieListItem
            mov={mov}
            key={mov.id}
            openDeleteModal={openDeleteModal || (() => {})}
          />
        ))
      ) : (
        <div css={[baphStyles.noResults]}>
          <h2>No Movies Match Your Search</h2>
          <div>Maybe you should go make one...</div>
        </div>
      )}
      <Button
        size={screenSize}
        onClick={navigateToArena}
        className={baphStyles}
      >
        Fight!
      </Button>
    </List>
  );
};

const baphStyles = {
  noResults: {
    display: 'flex',
    flexDirection: 'column' as 'column',
    alignItems: 'center',
    marginBottom: '30px',
    marginTop: '30px'
  },
  button: {
    alignSelf: 'end',
    display: 'flex',
    flexDirection: 'column' as 'column'
  }
};
