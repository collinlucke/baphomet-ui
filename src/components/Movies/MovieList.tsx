/** @jsxImportSource @emotion/react */
import {
  Button,
  List,
  Search,
  useResizedWidth
} from '@collinlucke/phantomartist';
import { MovieListItem } from './MovieListItem';
import { ChangeEvent, FormEventHandler } from 'react';
import { useNavigate } from 'react-router-dom';

type Movie = {
  id: string;
  title?: string;
  releaseDate?: string;
  rated?: string;
  poster?: string;
};

type MovieData = {
  movieData: {
    allMovies: Movie[];
  };
  searchTerm?: string;
  resultsCount?: number;
  useSearchButton?: boolean;

  onSearch?: React.FormEventHandler<HTMLFormElement>;
  setSearchTerm?: (e: ChangeEvent<HTMLInputElement>) => void;
};

export const MovieList: React.FC<MovieData> = ({
  movieData,
  searchTerm,
  useSearchButton,
  onSearch,
  setSearchTerm
}) => {
  const navigate = useNavigate();
  const width = useResizedWidth();

  const onSearchHandler: FormEventHandler<HTMLFormElement> = e => {
    onSearch?.(e);
  };

  const setSearchTermHandler = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchTerm?.(e);
  };

  const navigateToArena = () => {
    navigate('/arena');
  };

  const buttonSize = () => {
    if (width <= 580) {
      return 'medium';
    }
    return 'large';
  };

  return (
    <List>
      <Search
        onSearch={onSearchHandler}
        searchTerm={searchTerm}
        searchLabel="Search Movies"
        setSearchTerm={setSearchTermHandler}
        resultsCount={movieData.allMovies.length}
        buttonSize={buttonSize()}
        inputSize={buttonSize()}
        useSearchButton={useSearchButton}
      />
      {movieData.allMovies.length ? (
        movieData.allMovies.map(mov => <MovieListItem mov={mov} key={mov.id} />)
      ) : (
        <div css={[baphStyles.noResults]}>
          <h2>No Movies Match Your Search</h2>
          <div>Maybe you should go make one...</div>
        </div>
      )}
      <Button
        size={buttonSize()}
        onClick={navigateToArena}
        className={baphStyles}
      >
        Fight!
      </Button>
    </List>
  );
};

const baphStyles = {
  noResults: {
    display: 'flex',
    flexDirection: 'column' as 'column',
    alignItems: 'center',
    marginBottom: '30px',
    marginTop: '30px'
  },
  buttons: {
    alignSelf: 'end',
    display: 'flex',
    flexDirection: 'column' as 'column'
  }
};
