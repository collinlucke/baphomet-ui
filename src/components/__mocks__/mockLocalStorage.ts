export const mockLocalStorage = () => {
  localStorage.clear();
  const mockLocalStorage: Record<string, unknown> = {};
  global.localStorage = {
    length: 0,
    key: (index: number) => {
      const keys = Object.keys(mockLocalStorage);
      return keys[index] || null;
    },
    setItem: (key, value) => {
      mockLocalStorage[key] = value;
    },
    getItem: key => {
      const value = mockLocalStorage[key];
      return typeof value === 'string' ? value : null;
    },
    removeItem: key => {
      delete mockLocalStorage[key];
    },
    clear: () => {
      for (let key in mockLocalStorage) {
        delete mockLocalStorage[key];
      }
    }
  };
  return () => {};
};
