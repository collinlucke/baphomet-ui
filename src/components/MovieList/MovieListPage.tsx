import { useLayoutEffect } from 'react';
import { useQuery, useMutation, ApolloError } from '@apollo/client';
import { useNavigate } from 'react-router-dom';
import { GET_ALL_MOVIES } from '../../api/queries';
import { DELETE_MOVIE } from '../../api/mutations';
import { MovieList } from './MovieList';
import {
  Block,
  Button,
  ButtonGroup,
  InnerWidth,
  Modal
} from '@collinlucke/phantomartist';
import { ChangeEvent, useState, useEffect } from 'react';
import { useShowHeading, useError } from '../../contexts';
import { CustomErrorTypes } from '../../CustomTypes.types';

type Movie = {
  id: string;
  title: string;
  releaseDate?: string;
  rated?: string;
};

type MovieData = {
  allMovies: Movie[];
  searchTerm?: string;
};

export const MovieListPage = () => {
  const navigate = useNavigate();
  const [searchTerm, setSearchTerm] = useState<string>('');
  const [movieData, setMovieData] = useState<MovieData>({ allMovies: [] });
  const [movieToDelete, setMovieToDelete] = useState({ id: '', title: '' });
  const [isReady, setIsReady] = useState(false);
  const { showHeading, setShowHeading } = useShowHeading();
  const { error, setError } = useError();

  useEffect(() => {
    if (!showHeading) {
      setShowHeading(true);
    }
    setError(undefined);
    setMovieToDelete({ id: '', title: '' });
  }, [showHeading]);

  const { data, loading } = useQuery(GET_ALL_MOVIES, {
    variables: {
      limit: 100, // TODO: Hard coded until I get around to making a thingy to put put in a custom value
      searchTerm
    },
    fetchPolicy: 'network-only',
    nextFetchPolicy: 'cache-first',
    onCompleted: data => {
      setMovieData(data);
      setIsReady(true);
    }
  });

  useLayoutEffect(() => {
    if (!loading && !error && data) {
      setIsReady(true);
    }
  }, [loading, error, data]);

  const [deleteMovie] = useMutation(DELETE_MOVIE, {
    onCompleted: data => {
      if (data.deleteMovie) {
        navigate(0);
      }
    },
    onError: (error: ApolloError) => {
      setError(error as CustomErrorTypes);
    }
  });

  const setSearchTermHandler = (e: ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;
    setMovieData({ allMovies: [] });
    setSearchTerm(value);
  };

  const openDeleteModal = ({ id, title }: { id: string; title: string }) => {
    setMovieToDelete({ id, title });
  };

  const deleteMovieHandler = () => {
    deleteMovie({ variables: { id: movieToDelete.id } });
    setMovieToDelete({ id: '', title: '' });
  };

  return (
    <>
      {isReady && (
        <Block>
          <InnerWidth>
            <h2 css={baphStyles.h2}>Here's a List of Movies</h2>
            <MovieList
              movieData={movieData}
              searchTerm={searchTerm}
              setSearchTerm={setSearchTermHandler}
              openDeleteModal={openDeleteModal}
            />
          </InnerWidth>
          {movieToDelete.id && !error && (
            <Modal
              className={baphStyles}
              closeModal={() => setMovieToDelete({ id: '', title: '' })}
            >
              <h2 css={baphStyles.h2}>
                Are you sure you want to delete{' '}
                <span css={baphStyles.movieTitleToDelete}>
                  {movieToDelete.title}
                </span>
                ?
              </h2>
              <ButtonGroup>
                <Button onClick={deleteMovieHandler}>Delete</Button>
                <Button
                  kind="secondary"
                  onClick={() => setMovieToDelete({ id: '', title: '' })}
                >
                  Cancel
                </Button>
              </ButtonGroup>
            </Modal>
          )}
        </Block>
      )}
    </>
  );
};

const baphStyles = {
  h2: {
    marginBottom: '20px'
  },
  modal: {
    backgroundColor: `rgba(255,255,255,.75)`
  },
  movieTitleToDelete: {
    fontStyle: 'italic'
  }
};
