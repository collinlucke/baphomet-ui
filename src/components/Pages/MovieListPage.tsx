import { useState } from 'react';
import { useQuery } from '@apollo/client';
import { GET_ALL_MOVIES } from '../../api/queries';
import { MovieList } from '../Movies/MovieList';
import { Block, InnerWidth } from '@crazy-overlord/phantomartist';

type Movie = {
  id: string;
  title?: string;
  year?: number;
  rated?: string;
};

type MovieData = {
  getAllMovies: Movie[];
};

export const MovieListPage = () => {
  const [movieData, setMovieData] = useState<MovieData>({ getAllMovies: [] });

  const {} = useQuery(GET_ALL_MOVIES, {
    variables: {
      limit: 50 // TODO: Hard coded until I get around to making a thingy to put put in a custom value
    },
    onCompleted: data => {
      setMovieData(data);
    }
  });

  return (
    <>
      <Block>
        <InnerWidth>
          <MovieList movieData={movieData} />
        </InnerWidth>
      </Block>
    </>
  );
};
