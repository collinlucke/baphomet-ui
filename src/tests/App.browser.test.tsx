import { expect, describe, it, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import routes from '../routes';
import '@testing-library/jest-dom';
import { mockLocalStorage } from '../components/__mocks__/mockLocalStorage';
import {
  MemoryRouter,
  Route,
  Routes,
  createBrowserRouter,
  RouterProvider
} from 'react-router-dom';
import { MovieListPage } from '../Pages/MovieList/MovieListPage';
import { MockedProvider } from '@apollo/client/testing';
import { WelcomePage } from '../Pages/Welcome/WelcomePage';
import App from '../App';
import mockGetAllMovies from '../components/__mocks__/mockGetAllMovies';
import mockCheckAuth from '../components/__mocks__/mockCheckAuth';

beforeEach(() => mockLocalStorage());

const mocks = [mockGetAllMovies, mockCheckAuth];

describe('App', () => {
  describe(`'beenHereBefore' HAS NOT been set`, () => {
    it(`renders the <WelcomePage>`, async () => {
      const router = createBrowserRouter(routes);

      render(
        <MockedProvider>
          <RouterProvider router={router} />
        </MockedProvider>
      );

      expect(screen.getByTestId('welcome-page'));
    });

    it(`matches the snapshot with <WelcomePage>`, async () => {
      const router = createBrowserRouter(routes);

      const { container } = render(
        <MockedProvider>
          <RouterProvider router={router} />
        </MockedProvider>
      );

      expect(screen.getByTestId('welcome-page'));

      expect(container).toMatchSnapshot();
    });
  });

  describe(`'beenHereBefore' HAS been set`, () => {
    it(`redirects to MovieListPage`, async () => {
      localStorage.setItem('beenHereBefore', 'yup');

      render(
        <MockedProvider mocks={mocks}>
          <MemoryRouter>
            <App />
            <Routes>
              <Route path={'/'} element={<WelcomePage />} />
              <Route path={'/movielist'} element={<MovieListPage />} />
            </Routes>
          </MemoryRouter>
        </MockedProvider>
      );

      expect(await screen.findByTestId('movie-list'));
    });
  });
});
