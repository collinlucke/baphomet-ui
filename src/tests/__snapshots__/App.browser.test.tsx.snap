// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > 'beenHereBefore' HAS NOT been set > matches the snapshot with <WelcomePage> 1`] = `
<div>
  <div>
    <header
      class="pa-header css-o2h1fg"
      data-testid="main-page-heading"
    >
      <div
        class="pa-block css-xfdfty"
      >
        <div
          class="pa-inner-width css-l3e7vp"
        >
          <h1>
            <a
              data-testid="home-link"
              href="/"
            >
              Baphomet
            </a>
          </h1>
          <div>
            <button
              class="pa-button primary large css-z4rkrz"
              data-testid="add-new-movie-button"
            >
              <svg
                class=""
                color="currentColor"
                fill="none"
                height="17"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="3px"
                viewBox="0 0 24 24"
                width="17"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 4V20M20 12H4"
                  stroke="currentColor"
                />
              </svg>
              Add new movie
            </button>
          </div>
        </div>
      </div>
    </header>
    <main
      class="css-1vyba8i"
    >
      <div
        class="pa-block css-19umwiw"
        data-testid="welcome-page"
      >
        <header
          class="pa-header css-7czrpt"
        >
          <h1>
            WELCOME!!!
          </h1>
          Looks like you might be new around here...
        </header>
        <div
          class="css-yreaf4"
        >
          <img
            alt="Picture of a cuddly little demon"
            src="baphy-full-transparent.png"
          />
        </div>
        <div
          class="pa-inner-width css-zzctzn"
        >
          <p>
            If you've reached this page, Hello! I'm Collin Lucke, I'm a Frontend Software Engineer, and you've stumbled upon my little pet project, Baphomet. Don't worry, it's not an app that sends you to hell - that's TicTok. This guy, as of now, displays a list of movies (like, only a couple at the time of writing this) and you can click on one to get more info. As you'll see it's still very rough around the edges. VERY rough around the edges.
          </p>
          <p>
            On the frontend it's a React/Typescript Vite app using React Router to get around and styled using Emotion CSS. I'm also using a homegrown component library that I've regretfully name PhantomArtist (never tell yourself that no matter what, you'll call something by whatever the Wu Tang name generator gives you).
          </p>
          <p>
            The backend is comprised of an Apollo GraphQL Server calling into a MongoDB database and serving up date from a (soon to be expired) S3 bucket. Planning to move my storage locally.
          </p>
          <p>
            If you have a username/password, then you can login and play around with adding a movie or two. And if you'd like to peak at the code,
             
            <a
              href="http://github.com/collinlucke"
            >
              click this stuff
            </a>
             
          </p>
          <p>
            Towards the bottom of the list you'll see a "Fight!" button. That's for later.
          </p>
        </div>
        <button
          class="pa-button primary large css-1f24vce"
        >
          Have a Look Around!
        </button>
      </div>
    </main>
  </div>
</div>
`;
